fetchcode/package_registry/__init__.py

```diff
@@ -0,0 +1,123 @@
+import requests
+
+
+router = Router()
+
+
+class Data:
+    """
+    Represent the Data extracted from any URL/PURL
+    `homepage_url` : URL to the homepage for this package
+    `api_url` :  Return the package repository API URL to obtain structured data for this package
+    `vcs_url` : URL for VCS download
+    `codeview_url` : URL where the code can be browsed online
+    `bugtracking_url` : URL to the issue or bug tracker for this package
+    `download_url` :  Return the package repository download URL to download the actual archive of code of this package. 
+    `documentation_url` : URL where the documentation can be found for this package
+    `readme_url` : URL where readme can be found for this package
+    `tags` : List of tags for this package
+    `tags_download_urls` : List of download URLS for tags for this package
+    `tags_readme_urls` : List of readme URLS for tags for this package 
+    """
+
+    homepage_url = None
+    api_url = None
+    vcs_url = None
+    codeview_url = None
+    bugtracking_url = None
+    download_url = None
+    documentation_url = None
+    readme_url = None
+    tags = None
+    tags_download_urls = None
+    tags_readme_urls = None
+
+
+def info(url):
+    """
+    Return data according to the `url` string
+    `url` string can be purl too
+    """
+    if url:
+        try:
+            return router.process(url)
+        except NoRouteAvailable:
+            return
+
+
+def get_response(url):
+    resp = requests.get(url)
+    if "json" in dir(resp):
+        return resp.json()
+
+    raise Exception("Response of this URL does not contain json object")
+
+
+@router.route("pkg:cargo/.*")
+def cargo_data(purl):
+    """
+    Return `data` mapping of various URLs, from the `purl` string
+    """
+    purl_data = PackageURL.from_string(purl)
+    base_path = "https://crates.io"
+    name = purl_data.name
+    version = purl_data.version
+    api_url = f"{base_path}/api/v1/crates/{name}"
+    download_url = f"{api_url}/{version}/download"
+    readme_url = f"{api_url}/{version}/readme"
+    response = get_response(api_url)
+    crate = response.get("crate")
+    versions = response.get("versions", [])
+    homepage_url = None
+    documentation_url = None
+    codeview_url = None
+
+    if crate:
+        homepage_url = crate.get("homepage")
+        documentation_url = crate.get("documentation")
+        codeview_url = crate.get("repository")
+
+    tags = []
+    tags_download_urls = []
+    tags_readme_urls = []
+    for version in versions:
+        if "num" in version:
+            num = version.get("num")
+            tags.append(num)
+        if "dl_path" in version:
+            dl_path = version.get("dl_path")
+            tags_download_urls.append(f"{base_path}{dl_path}")
+        if "readme_path" in version:
+            readme_path = version.get("readme_path")
+            tags_readme_urls.append(f"{base_path}{readme_path}")
+
+    Data.homepage_url = homepage_url
```

Comment:
Data is not a super happy name... what is this Data about? Also you cannot use a class this way... you need to create an object... @sbs2001 can you show how to use dataclass and or attr to Tushar?

Reply:
Okay, got you, quite silly of me using class this way :p

Reply:
@pombredanne  re
>can you show how to use dataclass and or attr to Tushar?

There is probably some misunderstanding happening here on what  attr meant :) . 
@TG1999   `attr` was https://github.com/python-attrs/attrs. Your case will look like : 

```
 from attr import attrs, attrib 
 @attrs
 class Data:
   homepage_url = attrib()
   api_url = attrib()
  .....
  ....
```
You would then simply create a object like : 
```
my_data = Data(api_url=api_url,homepage_url=homepage_url .....)
```

@pombredanne  btw do tell which `attr` conventions you prefer :

For instance  you can do the same thing by 
```
import attr
@attr.s
class Data:
  api_url = attr.ib()
.......
```

Note the shorthand

Reply:
Thanks @sbs20001 😄